openshift-kube-apiserver/openshiftkubeapiserver/sdn_readyz_wait.go

```diff
@@ -0,0 +1,110 @@
+package openshiftkubeapiserver
+
+import (
+	gocontext "context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+func newOpenshiftAPIServiceReachabilityCheck() *openshiftAPIServiceAvailabilityCheck {
+	return &openshiftAPIServiceAvailabilityCheck{done: make(chan struct{})}
+}
+
+type openshiftAPIServiceAvailabilityCheck struct {
+	// done indicates that this check is complete (success or failure) and the check should return true
+	done chan struct{}
+}
+
+func (c *openshiftAPIServiceAvailabilityCheck) Name() string {
+	return "openshift-apiservices-available"
+}
+
+func (c *openshiftAPIServiceAvailabilityCheck) Check(req *http.Request) error {
+	select {
+	case <-c.done:
+		return nil
+	default:
+		return fmt.Errorf("check is not yet complete")
+	}
+}
+
+func (c *openshiftAPIServiceAvailabilityCheck) start(context genericapiserver.PostStartHookContext) error {
+	reachedOpenshiftAPIServer := make(chan struct{})
+	waitUntilCh := make(chan struct{})
+	defer func() {
+		close(waitUntilCh) // this stops the endpiont check
+		close(c.done)      // once this method is done, the ready check should return true
+	}()
+	start := time.Now()
+
+	kubeClient, err := kubernetes.NewForConfig(context.LoopbackClientConfig)
+	if err != nil {
+		return err
+	}
+
+	// start a thread which repeatedly tries to connect to any openshift-apiserver endpoint
+	go func() {
+		defer utilruntime.HandleCrash()
+
+		// we don't care about someone faking a response.
+		client := http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			},
+			Timeout: 1 * time.Second, // these should all be very fast.  if none work, we continue anyway.
+		}
+
+		wait.PollImmediateUntil(1*time.Second, func() (bool, error) {
+			ctx := gocontext.TODO()
+			openshiftEndpoints, err := kubeClient.CoreV1().Endpoints("openshift-apiserver").Get(ctx, "api", metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				// if we have no openshift apiserver endpoint, we have no reason to wait
+				klog.Warning("api.openshift-apiserver.svc endpoints were not found")
+				return true, nil
+			}
+			if err != nil {
+				utilruntime.HandleError(err)
+				return false, nil
+			}
+			for _, subset := range openshiftEndpoints.Subsets {
+				for _, address := range subset.Addresses {
+					resp, err := client.Get(fmt.Sprintf("https://%v:6443", address.IP))
```

Comment:
what if the port changes?

Reply:
> what if the port changes?

then this code will start producing warnings and we will see slow readyness. Also, don't do that.

Reply:
down here: https://github.com/openshift/kubernetes/pull/293/files#diff-c12df34becbda06f708eeb32ca6a43d5R107